Print more details on peers relying on ADCs.

ensonic opened the pull request:
This s a preparation to a policy on ADC usage.

ensonic commented:
New output is the part in ():
```
2023/03/03 17:13:48 Served access token to 172.26.0.39:36388 (ua="Fluent-Bit", pod="default/fluent-bit-7fbbc")
2023/03/03 17:13:48 Pod "gcscat" has no ip (yet): ""
2023/03/03 17:13:49 Retrying 1 pods without ip
2023/03/03 17:13:49 Retrying pod (kube-system/gcscat)
2023/03/03 17:13:49 Served access token to 172.26.0.252:41198 (ua="gcloud-golang/0.1", pod="kube-system/gcscat")
```
I'll eventually remove the 3 messages in the middle (retrying).

mp4096 commented on src/go/cmd/metadata-server/metadata.go:
```diff
@@ -181,12 +189,85 @@ func (th *TokenHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+			log.Printf("Failed to list pods in ns=%q: %v", nsName, err)
+		}
+		for _, pod := range pods.Items {
+			// log.Printf("Search pods for %q == %q (%s/%s)", ip, pod.Status.PodIP, nsName, pod.Name)
```
Remove?

mp4096 commented on src/go/cmd/metadata-server/metadata.go:
```diff
@@ -181,12 +189,85 @@ func (th *TokenHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		for _, p := range ptr {
+			nsName := p.ObjectMeta.Namespace
+			podName := p.ObjectMeta.Name
+			//log.Printf("Retrying pod (%s/%s)", nsName, podName)
```
Same

mp4096 commented on src/go/cmd/metadata-server/metadata.go:
```diff
@@ -135,6 +140,9 @@ func (th *TokenHandler) NewMetadataHandler(ctx context.Context) *MetadataHandler
 	}
 }
 
+// Return an access token for the 'robot-servce' service account
```
Typo? 
```suggestion
// Return an access token for the 'robot-service' service account
```

mp4096 commented on src/go/cmd/metadata-server/metadata.go:
```diff
@@ -181,12 +189,85 @@ func (th *TokenHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Collect some extra data for diagnostics (aka which services rely on ADCs).
+	// User-Agent: "Fluent-Bit", "gcloud-golang/0.1"
+	ua := r.Header.Get("User-Agent")
+	pod := th.getPodNameByIP(r.Context(), ipPort[0])
```
QQ: Assuming this gets really slow because we have many pods. How will it affect our system? Would it mean that some svcs will take longer to start up because it takes longer to get a token?

ensonic replied:
The http server uses go routines. So this will not as such cause a lot of extra load, but will incur some extra latency if a service asks for ADCs right away. I don't have a better idea for this yet.

mp4096 replied:
Yeah, I was not concerned about system load, more about the startup latency. Most of our services create ADC-based clients on startup and do it synchronously.

I think ~1 second latency should be ok, esp. given it should not stack for the overall startup time.

+1 for the flag.

ensonic commented:
I added some time measurements:
* when we have an IP already it completes quickly (~40ms) - this will get slower with more namespaces and pods, so some caching as per TOD in the change might be a good idea.
* when we don't have an IP yet, it took always at least 500 ms until get got an IP, but never longer than 1 s. We can play with the 2 constants I introduced (e.g. trying 250ms sleep) if we're concerned about the 1s delay. I think in practice many service might do something else before we get here.

I'll tweak the change to add a boolean flag that is false by default. With that we can submit this, roll it out and manually enable it for testing.

mp4096 commented on src/go/cmd/metadata-server/metadata.go:
```diff
@@ -30,6 +30,14 @@ import (
+)
+
+const (
+	GetPodByIPRetries = 10
```
does this need to be exported?

mp4096 commented on src/go/cmd/metadata-server/metadata.go:
```diff
@@ -30,6 +30,14 @@ import (
+
+const (
+	GetPodByIPRetries = 10
+	GetPodByIPWait    = 500 * time.Millisecond
```
same
